<table>
    <tr>
        <td><img src="https://www.acofi.edu.co/eiei2016/wp-content/uploads/2016/09/Logo-Universidad-EIA.jpg" width="250"/></td>
        <td>&nbsp;</td>
        <td>
        <td><img src="https://raw.githubusercontent.com/Fabian830348/Bases_Datos/refs/heads/master/Logo_EICT_horizontal_ESPANOL%20(1).png" width="300"/></td>
        <td>&nbsp;</td>
        <td>
            <h1 style="font-size:200%;color:blue;text-align:center">    <FONT COLOR="blue"> Regresión </p> Regresión Lineal </FONT>         </h1></td>         
        <td>
            <tp><p style="font-size:99%;text-align:center">Machine Learning </p></tp>
            <tp><p style="font-size:115%;text-align:center">A. Supervisado</p></tp>
            <tp><p style="font-size:115%;text-align:center">Prof. Fabián Sánchez</p></tp>
        </td>
    </tr>
</table>

# <FONT SIZE=5 COLOR="purple"> 1. Introducción </FONT>

**Modelo de regresión:** es el modelo en el que se busca establecer la relación entre un cierto número de características y una variable objetivo continua. Es decir, la relación entre la variable dependiente $y$ y las variables predictoras $x_1, x_2 , \dots , x_n$.

En esta sección trabajaremos el modelo de regresión lineal simple, es decir, univariado. Primero vamos a recordar algunos conceptos claves.

## <FONT SIZE=4 COLOR="green"> 1.1 Covarianza </FONT>

La covarianza es el valor que refleja en que medida dos variables aleatorias varian de forma conjunta respecto a sus medias.

De acuerdo con lo anterior, la covarianza está dada por la siguiente fórmula

$$Muestra: Cov(X,Y)=S_{xy}=\dfrac{\sum \limits_{i=1}^n (x_i-\overline{x})(y_i-\overline{y})}{n-1}$$

$$Poblacion:COV(X,Y)=\sigma_{XY}=\dfrac{\sum \limits_{i=1}^N (x_i-\mu_x)(y_i-\mu_y)}{N}$$

donde $\overline{x}$ y $\overline{y}$ son las medias de las variables $X$ y $Y$ respectivamente y $n$ es el total de observaciones.

A continuación listaremos algunas propiedades de la covarianza

- $Cov(X,c)=0$, donde $c$ es una constante.

- $Cov(X,X)=Var(X)$, la covarianza de una variable con si misma es la varianza de la variable.

- $Cov(X,Y)=C(Y,X)$, la varianza es simétrica, es decir, no importa el orden en que se coloquen la variables.

- La covarianza mide la fuerza de la relación lineal entre dos variables.

- Una alta covarianza no implica efecto causal.

**Interpretación de la covarianza**

- $Cov(X,Y)>0$ ; las variables $X$ y $Y$ tienden a moverse en la misma dirección.

- $Cov(X,Y)<0$   las variables $X$ y $Y$ tienden a moverse en direcciones opuestas.

- $Cov(X,Y)=0$: las variables $X$ y $Y$ no estan relacionadas linealmente.


## <FONT SIZE=4 COLOR="green"> 1.2 Coeficiente de Correlación </FONT>

El coeficiente de correlación se define como

$$muestra: r_{xy}=\dfrac{S_{xy}}{S_xS_y} \qquad \qquad poblacion : \rho_{xy}=\dfrac{\sigma_{xy}}{\sigma_{x}\sigma_{y}}$$

lo que es equivalente a

$$r_{xy}=\dfrac{\sum \limits_{i=1}^n (x_i-\overline{x})(y_i-\overline{y})}{\left (\sum \limits_{i=1}^n (x_i-\overline{x})^2 \right)^{1/2} \left ( \sum \limits_{i=1}^n (y_i-\overline{y})^2 \right)^{1/2}}$$

Veamos la interpretación de los diferentes valores del coeficiente de correlación de Pearson, teniendo en cuenta que:

$$-1 \leq r_{xy} \leq 1 $$

<FONT SIZE=3 COLOR="purple"> ***Punto para caso final: demostrar la desigualdad anterior*** </FONT>


- Si $r_{xy}=1$, hay una relación positiva perfecta, es decir, indica una dependencia total entre las variables de manera directa. Si una de ellas aumenta, la otra también lo hace en proporción constante. Es decir, la relación es lineal.

- Si $0<r_{x,y}<1$, existe una correlación positiva.

- Si $r_{x,y}=0$, no existe relación lineal. Pero pueden haber otro tipo de relaciones.

- Si $-1<r_{x,y}<0$, existe una correlación negativa.

- Si $r_{x,y}=-1$, existe una correlación negativa perfecta, es decir, indica una dependencia total de manera inversa. Si una variable disminuye la otra aumenta en proporción constante.

<br>
<center><img src="https://github.com/Fabian830348/cursos/blob/master/pearson.png?raw=true" alt="centered image" width="550" height="300"></center>
<br>

Algunas propiedades para tener en cuenta del coeficiente de correlación:

- No depende de las unidades de las variables.

- El coefiente de Pearson es acotado


# <FONT SIZE=5 COLOR="purple"> 2. Modelo de Regresión Lineal </FONT>

Supongamos que tenemos dos variables cuantitativas, que representan en el plano un conjunto de puntos $(x,y)$. En general, estos puntos no tienen porque estar ajustados de forma lineal, es decir, el diagrama de dispersión no tiene porque representar una recta.

Sin embargo, existen situaciones en que los puntos se aproximan o se ajustan en mayor medida a una línea recta. Particualmente cuando el coeficiente de correlación es cercano a 1 o a -1.

<br>
<center><img src="https://github.com/Fabian830348/cursos/blob/master/regresion1.png?raw=true" alt="centered image" width="550" height="300"></center>
<br>

La idea es encontrar una recta que se aproxime a los puntos dados

<br>
<center><img src="https://github.com/Fabian830348/cursos/blob/master/regresion2.png?raw=true" alt="centered image" width="550" height="300"></center>
<br>

Para hallar la ecuación de la recta que se observa en el gráfico se puede usar:

- Método de los mínimos cuadrados.

- Descenso del gradiente

<FONT SIZE=4 COLOR="blue"> Mínimos Cuadrados </FONT>

El método de los mínimos cuadrados permite hacer una estimación del modelo de regresión lineal simple.

Buscamos una recta de la forma

$$\hat{y}= \hat{\beta_0}+\hat{\beta_1}x$$

El método de mínimos cuadrados consiste en minimizar las sumas de los cuadrados de los errores $(SSE)$:

$$SSE=\sum \limits_{i=1}^n e_{i}^2= \sum \limits_{i=1}^n (y_{i}-\hat{y_i})^2$$

Esta expresión es la suma de las distancias al cuadrado entre los valores $y_i$ de la muestra (conjunto de datos) y los valores ajustados $\hat{y_i}$.

$$SSE(\hat{\beta_0},\hat{\beta_1})=\sum \limits_{i=1}^n (y_{i}-\hat{\beta_0}-\hat{\beta_1}x_i)^2$$

El objetivo de los mínimos cuadrados es hallar los valores de $\hat{\beta}_0$ y $\hat{\beta}_1$ que minimicen la expresión anterior. Para esto, realizaremos el procedimiento clásico para minimizar una función de dos variables y por tanto, derivamos parcialmente con respecto a los parámetros $\hat{\beta_0}$ y $\hat{\beta_1}$ e igualamos a cero.

$$\dfrac{\partial (SSE)}{\partial \hat{\beta}_0} = -2 \sum \limits_{i=1}^n (y_{i}-\hat{\beta}_0-\hat{\beta}_1x_i)$$
$$ \dfrac{\partial (SSE)}{\partial \hat{\beta_1}} = -2 \sum \limits_{i=1}^n (y_{i}-\hat{\beta_0}-\hat{\beta_1}x_i)x_i$$

el procedimiento que sigue se le deja al lector. Al igualar a cero y despejar $\hat{\beta_0}$ y $\hat{\beta_1}$ se obtiene:

$$\hat{\beta_0}= \dfrac{\sum \limits_{i=1}^n y_i - \hat{\beta}_1 \sum \limits_{i=1}^n x_i}{n}=\overline{y}-\hat{\beta_1}\overline{x} $$
$$\hat{\beta_1}=\dfrac{\sum \limits_{i=1}^n (x_i- \overline{x})(y_i-\overline{y})}{\sum \limits_{i=1}^n(x_i -\overline{x})^2}= \dfrac{S_{xy}}{S_{xx}}$$


donde, $S_{xy}$ es la covarianza muestral entre $X$ y $Y$.





Supongamos que tenemos los datos

In [ ]:
import numpy as np
# Creación de un conjunto de datos
X = [3,7,11,15,18,27,29,30,30,31,31,32,33,33,34,36,36,36,37,38,39,39,39,40,41,42,42,43,44,45,46,47,50]
Y = [5,11,21,16,16,28,27,25,35,30,40,32,34,32,34,37,38,34,36,38,37,36,45,39,41,40,44,37,44,46,46,49,51]

In [ ]:
import plotly.express as px
px.scatter(x = X, y = Y)

In [ ]:
# calcular la correlación de X , Y : usamos la librería de numpy
np.corrcoef(X,Y)[0,1]

In [ ]:
import plotly.express as px
# diagrama de dispersión
px.scatter(x = X,                                      # variable 1
           y = Y,                                      # variable 2
           color_discrete_sequence = ["purple"],       # color de los puntos
           trendline = "ols",                          # ols : ordinary least squares- mínimos cuadrados regresión lineal.
           trendline_color_override = "blue",          # color de la recta de regresión
           title = "Diagrama de dispersión con línea de regresión",
           labels = {"x" : "Metros Cuadrados", "y" : "Precio"})

Vamos a cargar las librerías necesarias para el trabajo de los ejemplos de regresión lineal

In [ ]:
#librerías para manipular dataframes y arreglos
import pandas   as pd
import numpy    as np

# librerías para graficar
import matplotlib       as plt
import plotly.express   as px
import seaborn          as sns

# selección de modelos y conjunto de entrenamiento
from sklearn.model_selection  import train_test_split

#métricas para modelo lineal
from sklearn.linear_model     import LinearRegression
from sklearn                  import linear_model

# métrica de evaluación
from sklearn.metrics          import mean_squared_error, r2_score

# warnings
import warnings
warnings.filterwarnings('ignore')


# <FONT SIZE=5 COLOR="purple"> 3. Ejemplo 1. Modelo de Regresión Lineal Simple </FONT>

Para este ejemplo consideraremos el conjunto de datos de apartamentos y las variables *precio* y *mt2*

In [ ]:
url = "https://raw.githubusercontent.com/Fabian830348/Bases_Datos/master/APARTAMENTOS.csv"
aptos = pd.read_csv(url)

Revisamos rápidamente algunos elementos de los datos

In [ ]:
aptos.head()

Unnamed: 0  precio    mt2 ubicacion  estrato  alcobas  banos balcon  \
0           1    79.0  43.16     norte        3        3      1     si   
1           2    93.0  56.92     norte        2        2      1     si   
2           3   100.0  66.40     norte        3        2      2     no   
3           4   123.0  61.85     norte        2        3      2     si   
4           5   135.0  89.80     norte        4        3      2     si   

  parqueadero  administracion     avaluo terminado  
0          si           0.050  14.923002        no  
1          si           0.069  27.000000        si  
2          no           0.000  15.738427        no  
3          si           0.130  27.000000        no  
4          no           0.000  39.567000        si

Eliminaremos la primera columna

In [ ]:
aptos.drop("Unnamed: 0" ,
           axis = 1 ,
           inplace = True)
aptos.head(5)

precio    mt2 ubicacion  estrato  alcobas  banos balcon parqueadero  \
0    79.0  43.16     norte        3        3      1     si          si   
1    93.0  56.92     norte        2        2      1     si          si   
2   100.0  66.40     norte        3        2      2     no          no   
3   123.0  61.85     norte        2        3      2     si          si   
4   135.0  89.80     norte        4        3      2     si          no   

   administracion     avaluo terminado  
0           0.050  14.923002        no  
1           0.069  27.000000        si  
2           0.000  15.738427        no  
3           0.130  27.000000        no  
4           0.000  39.567000        si

In [ ]:
# tamaño de los datos
aptos.shape

(694, 11)

Descripción estadística general de los datos

In [ ]:
aptos.describe()

precio         mt2     estrato     alcobas       banos  \
count   694.000000  694.000000  694.000000  694.000000  694.000000   
mean    317.762739  120.943818    4.651297    2.910663    2.289625   
std     247.614914   74.695493    1.208205    0.763962    0.795718   
min      25.000000   26.000000    2.000000    1.000000    1.000000   
25%     160.000000   71.000000    4.000000    3.000000    2.000000   
50%     245.000000   97.500000    5.000000    3.000000    2.000000   
75%     380.000000  141.750000    6.000000    3.000000    3.000000   
max    1700.000000  500.000000    6.000000   14.000000    6.000000   

       administracion       avaluo  
count      694.000000   694.000000  
mean         0.263556   182.636593  
std          0.251586   169.082384  
min          0.000000     0.149000  
25%          0.090000    73.000000  
50%          0.201000   131.431500  
75%          0.375000   234.051000  
max          2.280000  1540.620000

Alguna información general
  

In [ ]:
aptos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 694 entries, 0 to 693
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   precio          694 non-null    float64
 1   mt2             694 non-null    float64
 2   ubicacion       694 non-null    object 
 3   estrato         694 non-null    int64  
 4   alcobas         694 non-null    int64  
 5   banos           694 non-null    int64  
 6   balcon          694 non-null    object 
 7   parqueadero     694 non-null    object 
 8   administracion  694 non-null    float64
 9   avaluo          694 non-null    float64
 10  terminado       694 non-null    object 
dtypes: float64(4), int64(3), object(4)
memory usage: 59.8+ KB


Observe, que no tiene datos faltantes. Vamos a revisar la correlación de las variables cuantitativas de la base *aptos*.

- Para esto usamos el método  ***.corr( )***

In [ ]:
aptos.corr(numeric_only = True)

precio       mt2   estrato   alcobas     banos  \
precio          1.000000  0.858258  0.623480  0.186486  0.626714   
mt2             0.858258  1.000000  0.562502  0.309660  0.666821   
estrato         0.623480  0.562502  1.000000  0.067818  0.478938   
alcobas         0.186486  0.309660  0.067818  1.000000  0.348839   
banos           0.626714  0.666821  0.478938  0.348839  1.000000   
administracion  0.752635  0.774457  0.636884  0.162141  0.548519   
avaluo          0.789532  0.752844  0.590855  0.153776  0.525388   

                administracion    avaluo  
precio                0.752635  0.789532  
mt2                   0.774457  0.752844  
estrato               0.636884  0.590855  
alcobas               0.162141  0.153776  
banos                 0.548519  0.525388  
administracion        1.000000  0.699940  
avaluo                0.699940  1.000000

Observamos una fuerte correlación en las variables *precio* y *mt2*. Con coeficente de 0.858258

También podemos calcular la correlación de dos variables

In [ ]:
np.corrcoef(aptos.precio, aptos.mt2)[0,1]

np.float64(0.8582584650198594)

Ahora, revisaremos el diagrama de dispersión.

In [ ]:
px.scatter(x= aptos.mt2,
           y= aptos.precio,
           color_discrete_sequence = ["purple"],
           labels = {"x" : "mt2", "y" : "precio"})

Revisemos los nombres de las columnas del data.frame

In [ ]:
aptos.columns

Index(['precio', 'mt2', 'ubicacion', 'estrato', 'alcobas', 'banos', 'balcon',
       'parqueadero', 'administracion', 'avaluo', 'terminado'],
      dtype='object')

Para hacer el modelo, vamos a dividir los datos en variable de respuesta.

In [ ]:
# seleccionamos las variables (pero pilas que el modelo tiene lios con el tipo de elemento: series)
X = aptos.mt2
y = aptos.precio

In [ ]:
# variables predictoras y variable objetivo
X = pd.Series.to_frame(aptos.mt2)
y = pd.Series.to_frame(aptos.precio)

In [ ]:
# conjunto de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X,
                                                    y,
                                                    random_state = 1 ,
                                                    test_size = 0.2)

Generamos el modelo

In [ ]:
# modelo de regresión
lr = linear_model.LinearRegression()

In [ ]:
# entrenamos el modelo
lr.fit(X_train, y_train)

LinearRegression()

In [ ]:
# revisamos los coeficientes
print(lr.coef_)
print(lr.intercept_)

[[2.91202106]]
[-31.73264389]


# <FONT SIZE=5 COLOR="purple"> 4. Evaluación de un modelo de regresión </FONT>

En esta sección revisaremos la forma en que se evalúa un modelo de regresión.

***MSE : Error cuadrático medio***

Matemáticamente, el $MSE$ se puede calcular como la suma promedio de la diferencia al cuadrado entre el valor real y el valor previsto o estimado representado por el modelo de regresión (línea o plano).

$$MSE = \dfrac{1}{n} \sum \limits_{i=1}^{n} (y_i - \hat{y})^2 $$

<center><img src="https://github.com/Fabian830348/cursos/blob/f8b630f3ce112dc8d05c828d9b5d3ac328c9544f/Imagen/reg1.png?raw=true" alt="centered image" width="500" height="300"></center>

***$R^2$ coeficiente de determinación***

$R^2$ es la relación entre la regresión de suma de cuadrados (SSR) y la suma de cuadrados total (SST).

$$R^2 = \dfrac{SSR}{SST}= \dfrac{\sum \limits_{i=1}^{n} (\hat{y_i}-\overline{y})^2}{\sum \limits_{i=1}^{n} (y_i - \overline{y})^2}$$

<center><img src="https://github.com/Fabian830348/cursos/blob/f8b630f3ce112dc8d05c828d9b5d3ac328c9544f/Imagen/reg2.png?raw=true" alt="centered image" width="500" height="300"></center>

La regresión de suma de cuadrados (SSR) representa la variación total de todos los valores pronosticados encontrados en la línea o plano de regresión del valor medio de todos los valores de las variables de respuesta. La suma de los cuadrados totales $SST$ representa la variación total de los valores reales del valor medio de todos los valores de las variables de respuesta.

El valor $R^2$ se utiliza para medir la bondad de ajuste o la línea de mejor se ajusta. Cuanto mayor sea el valor de $R^2$, mejor será el modelo de regresión, ya que la mayor parte de la variación de los valores reales a partir del valor medio queda explicada por el modelo de regresión.

**NOTA**

En regresión lineal se recomienda utilizar $R^2$ para evaluar el rendimiento del modelo de los modelos de regresión. Esto se debe principalmente a que $R^2$  captura la fracción de varianza de los valores reales capturados por el modelo de regresión y tiende a brindar una mejor imagen de la calidad del modelo de regresión. Además, los valores de $MSE$ difieren en función de si los valores de la variable de respuesta están escalados o no. Una mejor medida en lugar de MSE es el error cuadrático medio (RMSE) que se ocupa del hecho relacionado con si los valores de la variable de respuesta están escalados o no.

**CONCLUSIONES**

- $MSE$ representa el error residual que no es más que la suma de la diferencia al cuadrado entre los valores reales y los valores previstos/estimados divididos por el número total de registros.

- $R^2$ representa la fracción de varianza capturada por el modelo de regresión.

- La desventaja de usar $MSE$ es que este valor varía en función de si los valores de la variable de respuesta están escalados o no. Si está escalado, el MSE será más bajo que los valores no escalados.

Vamos a calcular estos dos valores con el ejemplo anterior.

Veamos el $R^2$ y el $MSE$

In [ ]:
# entrenamiento
y_pred_train=lr.predict(X_train)
# prueba
y_pred_test=lr.predict(X_test)
print("R^2 entrenamiento", r2_score(y_train, y_pred_train))
print("R^2 prueba",  r2_score(y_test, y_pred_test))
print("MSE entrenamiento", mean_squared_error(y_train, y_pred_train))
print("MSE prueba", mean_squared_error(y_test, y_pred_test))

R^2 entrenamiento 0.7503603618113431
R^2 prueba 0.6648398911134308
MSE entrenamiento 15799.433033632618
MSE prueba 17591.433649590588


In [ ]:
# extraer la raíz RMSE
np.sqrt(15799.433)

np.float64(125.69579547462995)

# <FONT SIZE=5 COLOR="purple"> 5. Ejemplo 2. Modelo de Regresión Lineal Múltiple</FONT>

Vamos a considerar los siguientes datos relacionados con la inversión en publicidad en diferentes medios y las ventas.

In [ ]:
ventas = pd.read_csv("https://raw.githubusercontent.com/Fabian830348/Bases_Datos/refs/heads/master/Ventas.csv", sep = ";" , decimal = "," )
ventas

Primero, vamos a explorar rápidamente los datos

In [ ]:
# tamaño de los datos
ventas.shape

In [ ]:
# nombre de las variables
ventas.columns

In [ ]:
# información de los datos
ventas.info()

In [ ]:
# estadísticas básicas
ventas.describe()

Vamos a eliminar la primera columna que no aporta información

In [ ]:
# eliminar la primera columna
ventas = ventas.drop(ventas.columns[0], axis=1)

In [ ]:
ventas.head()

Grafíquemos los diagramas de dispersión

In [ ]:
sns.pairplot(ventas,
             x_vars=['Newspaper','TV','Radio', 'Web'],
             y_vars='Sales',
             size=7,
             aspect=0.8,
             kind = 'reg')

calculamos las correlaciones de las variables

In [ ]:
corr = ventas.corr()
corr

In [ ]:
# matriz de correlación
sns.heatmap(corr, annot = True,
            yticklabels=corr.columns,
            xticklabels=corr.columns)

De los gráficos de dispersión y la matriz de correlación se puede identificar cuáles variables guardan una mayor relación con las ventas. Por ejemplo, *newspaper* tal vez tenga poco impacto en las ventas esto por la alta dispersión de los datos. Pero *TV* tiene un fuerte incidencia.

In [ ]:
# variable objetivo
y = ventas['Sales']
# variables predictoras
X = ventas.drop('Sales', axis=1)
# conjunto de entrenamiento y prueba
X_train,X_test,y_train,y_test=train_test_split(X,
                                               y,
                                               train_size=.70,
                                               random_state=123)
# definimos el modelo
modelo = LinearRegression()
# entrenamo el modelo
modelo.fit(X_train,y_train)

In [ ]:
print(modelo.intercept_)
print(modelo.coef_)

Esto indica que:

$Sales = 2.34+0.003*Newspaper+ 0.045*TV+0.18*Radio+0.002*Web$

In [ ]:
# predicciones en entrenamiento
y_pred_train=modelo.predict(X_train)
# predicciones en prueba
y_pred_test=modelo.predict(X_test)
print("R2 entrenamiento:",r2_score(y_train, y_pred_train))
print("R2 prueba:",r2_score(y_test, y_pred_test))
print("MSE entrenamiento:", mean_squared_error(y_train, y_pred_train))
print("MSE prueba:",  mean_squared_error(y_test, y_pred_test))
print("RMSE entrenamiento:", np.sqrt(mean_squared_error(y_train, y_pred_train)))
print("RMSE prueba:",  np.sqrt(mean_squared_error(y_test, y_pred_test)))

# <FONT SIZE=5 COLOR="purple"> 6. Ejemplo 3. Modelo de Regresión Lineal Simple </FONT>

Importamos la base de los pingüinos

In [ ]:
df = sns.load_dataset("penguins")
df

Revisamos si hay datos nulos

In [ ]:
df.isnull().sum()


Los eliminamos

In [ ]:
df.dropna(inplace=True)

**Revisamos relaciones lineales**

Con la función pairplot podemos rápidamente visualizar la distribución de nuestros datos y cruzar datos en gráficos de dispersión.

Los parámetros son:

- ***hue***: con el cual podemos dividir los datapoints según etiquetas (en este caso especies de pingüinos).

- ***markers***: que son una leyenda para identificar más fácilmente nuestros puntos en la gráfica.

In [ ]:
sns.pairplot(df, hue="species", markers=["o", "s", "D"])

Revisamos la correlación

In [ ]:
corr = df.corr(numeric_only=True)
corr

In [ ]:
# matriz de correlación
sns.heatmap(corr, annot = True, yticklabels=corr.columns, xticklabels=corr.columns)

Hacemos un gráfico de correlación lineal.

En este caso suponemos que existe una relación entre la masa corporal de los pingüinos y el tamaño de sus alas.

In [ ]:
sns.regplot(x=df["body_mass_g"],
            y=df["flipper_length_mm"],
            color = "blue")

**Definimos el modelo**

- A partir de la correlación linear anterior, entrenaremos un modelo simple de regresión lineal.

Lo primero es crear dos vectores. En este caso usamos la masa corporal como nuestro vector X (predictor) y el tamaño de las alas como el vector y (valor por predecirse).

La función **train_test_split**  nos divide nuestro data sets en datos de entrenamiento y prueba. Los datos de prueba nos servirán para medir el desempeño de nuestro modelo.

Las funciones *values* y *reshape* nos permiten crear un arreglo artificial en 2D que nos permitirá entrenar nuestro modelo.

Para entrenar al modelo, solo tenemos que invocar a Linear Regression y luego entrenar al modelo con la función fit.

In [ ]:
X = df["body_mass_g"].values.reshape(-1,1)
y = df["flipper_length_mm"]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 0)

In [ ]:
reglin = LinearRegression()
reglin.fit(X_train, y_train)

**Suma de Errores Cuadráticos**

En este modelo tenemos ya un modelo entrenado y listo para ser medido.

Lo que tenemos es una vector llamado y_test (datos reales) que utiliza a X_test para generar una predicción. Lo que hacemos con una sola línea de código es sumar todos los vectores al cuadrado.

In [ ]:
sec = np.sum((y_test - reglin.predict(X_test))**2)
sec

Si vemos el promedio

In [ ]:
((y_test - reglin.predict(X_test))**2).mean()

**Evaluación del modelo**

En este caso usamos dos métricas adicionales que vienen preconstruidas en Scikit Learn.

Por un lado tenemos al promedio de los errores cuadráticos (mean_squared_error). Como su nombre lo indica, hace un promedio de los errores cuadráticos en vez de sumarlos.

Una métrica que se usa popularmente en regresión linear son los mínimos cuadrados ordinarios $R^2$. Este genera un rango de 0 a 1 para medir la varianza capturada por el modelo.

In [ ]:
y_test_predict = reglin.predict(X_test)

rmse = mean_squared_error(y_test, y_test_predict)
r2 = r2_score(y_test, y_test_predict)

print(rmse)
print(r2)

# Observación Gráfica en tres dimensiones

Cuando trabajamos con tres dimensiones, podemos visualizar nuestros datos usando un cubo. Para esto, importaremos las siguientes librerías y módulos

***mplot3d*** y ***Axes3D*** de *mpl_toolkits*: con los que podemos hacer gráficas en tres dimensiones.




In [ ]:
data = np.array(df[["bill_length_mm","flipper_length_mm", "body_mass_g"]])
print(data.shape)

Visualizar en **3D**

In [ ]:
# Import libraries
from mpl_toolkits import mplot3d
import numpy as np
import matplotlib.pyplot as plt
import scipy.linalg


# Variables
z = df["bill_length_mm"]
x = df["flipper_length_mm"]
y = df["body_mass_g"]

# Creando la figura
fig = plt.figure(figsize = (10, 7))
ax = plt.axes(projection ="3d")

# Creando el gráfico
ax.scatter3D(x, y, z, color = "red")
plt.title("simple 3D scatter plot")

# Mostrar la gráfica
plt.show()

In [ ]:
z = df["bill_length_mm"]
x = df["flipper_length_mm"]
y = df["body_mass_g"]
fig = px.scatter_3d(df, x, y, z, color = "species" )
fig.show()

# Limitaciones de la regresión Lineal

Nos enfocaremos en dos de las limitantes más importantes de regresión lineal: outliers y heterocedasticidad.

In [ ]:
prop = sns.load_dataset("tips")
prop

No usaremos variables categóricas

In [ ]:
prop.drop(["sex", "smoker", "day", "time"], axis=1, inplace=True)

Las columnas restantes son:

total_bill: el total de la cuenta
tip: propinas
size: número de comensales

## Outliers

Poder identificar a los outliers es muy sencillo. Todo lo que tenemos que hacer es utilizar la función boxplot dentro de seaborn para generar un diagrama de caja.

In [ ]:
%matplotlib inline
plt.rcParams['figure.figsize'] = (20, 10)
plt.style.use('ggplot')

ax = sns.boxplot(x=prop["total_bill"])

Lo que podemos apreciar es el

- Mínimo y máximo no atípico, representados por las líneas en los extremos.

- El rango intercuartílico, representado por la caja en rojo. En este caso, tenemos un rango que va del cuartil 1 (Q1) al cuartil 3 (Q3). Dentro de este encontramos a la mayoría de nuestros puntos.

- La línea dentro de la caja roja que representa a la mediana.

- Los puntos fuera del máximo no atípico, que son los outliers o valores atípicos.

- En este caso podemos ver que tenemos 8 valores atípicos que podría afectar a nuestro modelo.

## Heterocedasticidad

Para evaluar la heterocedasticidad, solo tenemos que correr de nuevo la función regplot dentro de Seaborn.

En este caso evaluaremos la correlación entre el total de la cuenta y las propinas.

In [ ]:
sns.regplot(x=prop["total_bill"], y=prop["tip"])

- En este caso podemos ver hay más homocedasticidad (lo contrario de heterocedasticidad) al principio de la línea, pero que la varianza aumenta al avanzar la línea.

- Es decir, las propinas suelen ser similares cuando la cuenta es pequeña. Sin embargo, varía más a medida que la cuenta se vuelve más grande.

In [ ]:
from sklearn.linear_model import LinearRegression, Ridge, Lasso

In [ ]:
prop.dropna(inplace=True)

In [ ]:
X = prop["total_bill"].values.reshape(-1,1)
y = prop["tip"]

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 42)

# Lineal vs Ridge vs Lasso

En esta parte solo haremos el entrenamiento de estas tres regresiones.

Los métodos de subset selection descritos anteriormente emplean mínimos cuadrados ordinarios (OLS) para ajustar un modelo lineal que contiene únicamente un subconjunto de predictores. Otra alternativa, conocida como regularización o shrinkage, consiste en ajustar el modelo incluyendo todos los predictores pero aplicando una penalización que fuerce a que las estimaciones de los coeficientes de regresión tiendan a cero. Con esto se intenta evitar overfitting, reducir varianza, atenuar el efecto de la correlación entre predictores y minimizar la influencia en el modelo de los predictores menos relevantes. Por lo general, aplicando regularización se consigue modelos con mayor poder predictivo (generalización). Tres de los métodos de regularización más empleados son Ridge, Lasso y Elastic net.

Dado que estos métodos de regularización actúan sobre la magnitud de los coeficientes del modelo, todos deben de estár en la misma escala, por esta razón es necesario estandarizar o normalizar los predictores antes de entrenar el modelo. Ls métodos están especialmente indicados para situaciones en las que hay un mayor número de predictores que de observaciones.

## Ridge

- Penaliza la suma de los coeficientes elevados al cuadrado

$$\|\beta\|^2 = \sum \limits_{j=1}^{p} \beta_j^2$$

$$\text{Suma residuos cuadrado} + \lambda \sum_{j=1}^p \beta_j ^2$$

- La principal ventaja de aplicar ridge frente al ajuste por mínimos cuadrados ordinarios (OLS) es la reducción de varianza.

- Desventaja es que tiene en cuenta todos los predictores

## Lasso

- - Penaliza la suma de los coeficientes elevados al cuadrado

$$\|\beta\|_1 = \sum \limits_{j=1}^{p} |\beta_j|$$

$$\text{Suma residuos cuadrado} + \lambda \sum \limits_{j=1}^{p} |\beta_j|$$

- La principal ventaja de aplicar lasso es que consigue excluir los predictores menos relevantes.

- Lasso es inestable si los predictores están altamente correlacionados. Es decir, si cambio los datos de entrenamientos existen cambios significativos.



In [ ]:
reg = LinearRegression()
reg.fit(X_train, y_train)

In [ ]:
lasso = Lasso()
lasso.fit(X_train, y_train)

In [ ]:
ridge = Ridge()
ridge.fit(X_train,y_train)

In [ ]:
plt.plot(X_test, reg.predict(X_test))
plt.plot(X_test, lasso.predict(X_test))
plt.plot(X_test, ridge.predict(X_test))
plt.legend(["Linear", "Lasso", "Ridge"])
plt.xlabel("Total Cuenta")
plt.ylabel("Propina")

plt.show()

# Evaluación de Modelos

Cuando trabajamos con mahcine learning, algo común es entrenar varios modelos y luego medir sus desempeños para decidir cual de ellos usar.

En este caso solo mediremos el desempeño de la regresión linear y lasso, usando el promedio de errores cuadrados y $R^2$ (coeficiente de determinación)

In [ ]:
y_reg_predict = reg.predict(X_test)
y_lasso_predict = lasso.predict(X_test)

In [ ]:
rmse = mean_squared_error(y_test, y_reg_predict)
r2 = r2_score(y_test, y_reg_predict)

print(rmse)
print(r2)

In [ ]:
rmse = mean_squared_error(y_test, y_lasso_predict)
r2 = r2_score(y_test, y_lasso_predict)

print(rmse)
print(r2)

Podemos graficar los errores residuales

In [ ]:
plt.style.use('fivethirtyeight')
plt.scatter(reg.predict(X_train), reg.predict(X_train) - y_train,
            color = "red", s = 10, label = 'Datos de Entrenamiento')
plt.scatter(reg.predict(X_test), reg.predict(X_test) - y_test,
            color = "blue", s = 10, label = 'Datos de Prueba')
plt.hlines(y = 0, xmin = 0, xmax = 7, linewidth = 2)
plt.legend(loc = 'upper right')
plt.title("Errores Residuales de Regresión Lineal")

plt.show()

In [ ]:
plt.style.use('fivethirtyeight')
plt.scatter(lasso.predict(X_train), lasso.predict(X_train) - y_train,
            color = "red", s = 10, label = 'Datos de Entrenamiento')
plt.scatter(lasso.predict(X_test), lasso.predict(X_test) - y_test,
            color = "blue", s = 10, label = 'Datos de Testeo')
plt.hlines(y = 0, xmin = 0, xmax = 7, linewidth = 2)
plt.legend(loc = 'upper right')
plt.title("Errores Residuales de Regresión Lasso")

plt.show()